README.md
# Dear Diary Readme

## Description

**Dear Diary** is a lightweight digital diary which is client - server based 

It's optimized for a PC with a Microsoft Edge or a Google Chrome Browser but i'ts possible to use it also on a Smartphone's browser (MS Edge or Google Chrome recommended).
___


CHANGELOG.md
# Release Notes

## V0.1.0 (First test version) -> 30.Jul 2022
- Login/Logout and register page created
- Dahboard created
- Creating new entries is possible

## V0.2.0 (Test version) -> 11.Oct 2022
- Refined design & responsiveness
- Creating new entries (Tagging functionality added)
- Editing and deleting memories

## V0.3.0 (test version) -> 06. Nov 2022
- Filter possibility implemented

## V0.4.0 (test version) -> 08. Dec 2022
- Image upload
- Detail view for a memory

## V1.0.0 (First live version)-> 11. Dec 2022
- Implemented some smalle code optimations


src/server.js
"use strict";

// Module einbinden
const crypto = require("crypto");
const fs = require("fs");
const http = require("http");
const { getMemories, insertMemory, deleteMemory } = require("./api/memory.js");
const {
  getBody,
  getBodyParams,
  getBodyJSON,
  getCookies,
  executeSnippets,
} = require("./lib/helpers.js");

// Statische Daten einbinden
const accesslist = require("./config/accesslist.json");
const config = require("./config/config.json");
const mimetypes = require("./data/mimetypes.json");
const users = require("./data/users.json");

// Templates
const templates = {};
for (const file of fs.readdirSync('./templates')) {
  if (!file.endsWith('.html')) continue;
  templates[file.split('.')[0]] = fs.readFileSync(`./templates/${file}`, 'utf8');
}

// Runtime
const loginSecurity = {
  attempts: {},
  timeouts: {},
};
const sessions = Object.create(null);
const sessionTimeouts = {};

// Webserver erstellen
const server = http.createServer(function (request, response) {
  // URL der Anfrage vorbereiten und direkt im Request wieder speichern
  request.url = new URL(request.url, "http://localhost");

  /*******************************************/
  /* SESSION WIEDERHERSTELLEN UND VERLÄNGERN */
  /*******************************************/

  // Session des Benutzers abfragen (falls sie existiert)
  const { sessionID } = getCookies(request);
  const session = sessions[sessionID];

  // Wenn eine Session existiert, dann verlängern wir diese um einen weiteren Gültigkeitszeitraum
  if (session) {
    // Aktuell laufenden Timeout zum automatischen Löschen der Session stoppen
    clearTimeout(sessionTimeouts[session.username]);

    // Einen neuen Timeout zum automatischen Löschen der Session starten
    sessionTimeouts[session.username] = setTimeout(function () {
      delete sessions[sessionID];
    }, 60 * 1000 * config.session.duration);

    // Session Cookie mit der verlängerten Ablaufdauer setzen
    response.setHeader(
      "Set-Cookie",
      `sessionID=${sessionID}; Max-Age=${
        60 * config.session.duration
      }; HttpOnly`
    );
  }

  /*************************************************/
  /* ENDPUNKTE FÜR LOGIN, LOGOUT UND REGISTRIERUNG */
  /*************************************************/

  //
  // Login Endpunkt
  // POST /?login
  if (
    request.url.pathname === "/" &&
    request.url.searchParams.get("login") === "" &&
    request.method === "POST"
  ) {
    // Eingehenden Body als Query String sammeln und verarbeiten
    getBodyParams(request, function (error, params) {
      // Fehler beim Sammeln des Body mit dem Status Code 500 (Internal Server Error) beantworten
      if (error) {
        response.endWithStatus(500);
        return;
      }

      // Prüfen, dass die gesendeten Daten einen Benutzernamen sowie ein Passwort enthalten sowie
      // innerhalb der Zeichenlimits für Zugangsdaten sind. Bei Problemen antworten wir mit einem
      // Status Code 400 (Bad Request)
      if (
        !params.username ||
        params.username.length < config.users.limits.username.min ||
        params.username.length > config.users.limits.username.max ||
        !params.password ||
        params.password.length < config.users.limits.password.min ||
        params.password.length > config.users.limits.password.max
      ) {
        response.endWithStatus(400);
        return;
      }

      // Benutzername und Passwort aus dem geparsten Body der Anfrage extrahieren
      let { username, password } = params;

      // abort right away if there is still a login timeout running for the user
      if (loginSecurity.attempts[username] === false) {
        response.endWithStatus(425);
        return;
      }

      // Das Passwort hashen, sodass wir es mit dem gespeicherten Passwort vergleichen können. Der
      // Hash setzt sich zusammen aus dem Benutzernamen, seinem Passwort sowie dem globalen Hash,
      // sodass daraus keine Rückschlüsse durch Angriffe Rainbow Tables gewonnen werden können.
      password = crypto
        .createHash(config.crypto.hash)
        .update(`${username}:${password}:${config.crypto.salt}`)
        .digest("hex");

      // Prüfen, ob der Benutzer nicht existiert oder das Passwort falsch ist und mit dem Status
      // Code 403 (Forbidden) antworten
      if (!users[username] || users[username].password !== password) {
        // Counter für ungültige Anmeldeversuche sicherstellen und dessen Wert erhöhen
        loginSecurity.attempts[username] ??= 0;
        loginSecurity.attempts[username]++;

        // Prüfen, ob der Benutzer mehr ungültige Anmeldeversuche als erlaubt unternommen hat und
        // falls ja, dann deaktivieren wir den Login komplett
        if (
          loginSecurity.attempts[username] >= config.users.limits.login.attempts
        ) {
          loginSecurity.attempts[username] = false;
        }

        // Bestehende Timeouts zum Aufräumen stoppen und einen neuen Timeout zum Zurücksetzen des
        // Login-Zählers bzw. der Anmeldesperre setzen
        clearTimeout(loginSecurity.timeouts[username]);
        loginSecurity.timeouts[username] = setTimeout(function () {
          // Zähler und Timeout für den Benutzer zurücksetzen, wodurch der Login wieder erlaubt wird
          delete loginSecurity.attempts[username];
          delete loginSecurity.timeouts[username];
        }, config.users.limits.login.block * 1000 * 60);

        // Gescheiterten Anmeldeversuch mit dem Status Code 403 (Forbidden) beantworten
        response.endWithStatus(403);
        return;
      }

      // An dieser Stelle angelangt wissen wir nun, dass ein registrierter Benutzer sich mit seinem
      // korrekten Passwort am Server angemeldet hat

      // Gibt es bereits eine Session, allerdings für einen anderen Benutzer, dann melden wir diesen
      // zunächst ab
      if (session && session.username !== username) {
        // Timeout zum automatischen Ablaufen der Session löschen
        clearTimeout(sessionTimeouts[username]);
        delete sessionTimeouts[username];

        // Session Inhalt löschen
        delete sessions[sessionID];
      }

      // Sollte der Benutzer noch keine Session haben, weil er noch nicht angemeldet ist, dann
      // erstellen wir nun eine neue Session inklusive Cookie
      if (!sessions[sessionID]) {
        // Session ID zufällig generieren
        const sessionID = crypto
          .randomBytes(config.session.signs / 2)
          .toString("hex");

        // Benutzerprofil aus der Users Datenbank in die Session mit der generierten ID speichern
        sessions[sessionID] = users[username];

        // Timeout zum automatischen Ablaufen der Session erstellen
        sessionTimeouts[username] = setTimeout(function () {
          delete sessions[sessionID];
        }, 60 * 1000 * config.session.duration);

        // Session ID mittels Cookie, welches ein Ablaufdatum hat sowie nicht per JavaScript ausge-
        // lesen werden darf, an den Client senden
        response.setHeader(
          "Set-Cookie",
          `sessionID=${sessionID}; Max-Age=${
            60 * config.session.duration
          }; HttpOnly`
        );
      }

      // Anfrage des Benutzers an die Startseite nach dem Login weiterleiten
      response.statusCode = 302;
      response.setHeader("Location", config.session.homepage);

      // Anfrage abschließen
      response.end();
    });

    // Dieser Endpunkt ist eine in sich geschlossene Einheit und die Anfrage wurde beantwortet
    return;
  }

  //
  // Logout Endpunkt
  // GET /?logout
  if (
    request.url.pathname === "/" &&
    request.url.searchParams.get("logout") === "" &&
    request.method === "GET"
  ) {
    // Nur wenn der Benutzer eine Session hat diese Löschen, was zB nicht der Fall ist, wenn ein
    // nicht angemeldeter Benutzer den Logout Endpunkt ansteuert
    if (session) {
      // Timeout zum automatischen Ablaufen der Session löschen
      clearTimeout(sessionTimeouts[session.username]);
      delete sessionTimeouts[session.username];

      // Session Inhalt löschen
      delete sessions[sessionID];
    }

    // Dem Client mitteilen, dass dieser sein Session Cookie löschen soll, indem wir die Lebensdauer
    // des Cookie auf 0 setzen
    response.setHeader("Set-Cookie", "sessionID=; Max-Age=0; HttpOnly");

    // Anfrage des Benutzers nach dem Logout zur Anmeldeseite weiterleiten
    response.statusCode = 302;
    response.setHeader("Location", "/");
    response.end();

    // Dieser Endpunkt ist eine in sich geschlossene Einheit und die Anfrage wurde beantwortet
    return;
  }

  //
  // Register Endpunkt
  // POST /?register
  if (
    request.url.pathname === "/" &&
    request.url.searchParams.get("register") === "" &&
    request.method === "POST"
  ) {
    // Eingehenden Body als Query String sammeln und verarbeiten
    getBodyParams(request, function (error, params) {
      // Fehler beim Sammeln des Body mit dem Status Code 500 (Internal Server Error) beantworten
      if (error) {
        response.endWithStatus(500);
        return;
      }

      // Prüfen, dass die gesendeten Daten einen Benutzernamen sowie ein Passwort enthalten sowie
      // innerhalb der Zeichenlimits für Zugangsdaten sind. Bei Problemen antworten wir mit einem
      // Status Code 400 (Bad Request) antworten
      if (
        !params.username ||
        params.username.length < config.users.limits.username.min ||
        params.username.length > config.users.limits.username.max ||
        !params.password ||
        params.password.length < config.users.limits.password.min ||
        params.password.length > config.users.limits.password.max
      ) {
        response.endWithStatus(400);
        return;
      }

      // Benutzername und Passwort aus dem geparsten Body der Anfrage extrahieren
      let { username, password } = params;

      // Wenn der Benutzer bereits existiert mit dem Status Code 409 (Conflict) antworten
      if (users[username]) {
        response.endWithStatus(409);
        return;
      }

      // Neue zufällige ID für den Benutzer generieren
      const id = crypto.randomUUID();

      // Das Passwort hashen, da wir niemals Passwörter im Klartext auf der Server speichern! Der
      // Hash setzt sich zusammen aus dem Benutzernamen, seinem Passwort sowie dem globalen Hash,
      // sodass daraus keine Rückschlüsse durch Angriffe Rainbow Tables gewonnen werden können.
      password = crypto
        .createHash(config.crypto.hash)
        .update(`${username}:${password}:${config.crypto.salt}`)
        .digest("hex");

      // Neuen Eintrag für den zu registrierenden Benutzer in der Users Datenbank erstellen
      users[username] = {
        id,
        groups: [],
        username,
        password,
      };

      // Die Users Datenbank speichern und dem Benutzer entsprechend antworten
      fs.writeFile(
        "data/users.json",
        JSON.stringify(users, null, 2),
        function (error) {
          // Wenn ein Fehler beim Speichern der Datenbank auftritt dem Benutzer mit dem Status Code
          // 500 (Internal Server Error) antworten
          if (error) {
            response.endWithStatus(500);
            return;
          }

          // Die erfolgreiche Registrierung mit dem Status Code 302 beantworten und
          // Den User an die Index Seite weiterleiten
          response.statusCode = 302;
          response.setHeader("Location", "index.html");

          // Anfrage abschließen
          response.end();
          return;
        }
      );
    });

    // Dieser Endpunkt ist eine in sich geschlossene Einheit und die Anfrage wurde beantwortet
    return;
  }

  /*********************/
  /* WEITERE ENDPUNKTE */
  /*********************/

  // Eigene Endpunkte hier einbinden

  /***********************/
  /* SQLITE DB ENDPOINTS */
  /***********************/

  // Get memory entries according to the query string
  if (
    session &&
    request.url.pathname === "/api" &&
    request.url.searchParams.get("getMemories") === "" &&
    request.method === "GET"
  ){
    const userId = sessions[sessionID].id;
    const params = {};

    // Create a key - value array pair and send it to the memory.js api
    for(let entry of request.url.searchParams.entries()) {
      params[entry[0]] = entry[1];
    }

    getMemories(userId, params, (err, resData) => {
      if (err){
        response.endWithStatus(400);
        return;
      }
      response.end(JSON.stringify(resData));
      return;
    })
    return;
  }

  // Insert new memory entry
  if (
    session &&
    request.url.pathname === "/api" &&
    request.url.searchParams.get("setMemory") === "" &&
    request.method === "POST"
  ){
    getBodyParams(request, function (error, params) {
      const userId = sessions[sessionID].id;
      if (error){
        response.endWithStatus(500);
        return;
      }

      // Insert the extracted body params to the sqlite db
      insertMemory(userId, params, (err, uuid) => {
        if (err){
          response.endWithStatus(400);
          return;
        }
        response.end(uuid);
        return;
      });
      return;
    });
    return;
  }

   // Delete memory entry
   if (
    session &&
    request.url.pathname === "/api" &&
    request.url.searchParams.get("deleteMemory") === "" &&
    request.method === "POST"
  ){

    getBodyParams(request, function (error, params) {
      if (error){
        response.endWithStatus(500);
        return;
      }

      const userId = sessions[sessionID].id;

      // Insert the extracted body params to the sqlite db
      deleteMemory(userId, params, (err, uuid) => {
        if (err){
          response.endWithStatus(400);
          return;
        }
        response.end(uuid);
        return;
      });
      return;
    });
    return;
  }

  /*******************************************************/
  /* HTTP METHODE PRÜFEN UND INDEX.HTML DATEIEN ERGÄNZEN */
  /*******************************************************/

  // Da die Endpunkte nun bearbeitet wurden sind wir im Bereich des reinen Fileservers, welcher nur
  // GET und HEAD Anfragen beantwortet. Alle anderen Anfragen werden explizit nicht unterstützt.
  if (request.method !== "GET" && request.method !== "HEAD") {
    response.endWithStatus(405);
    return;
  }

  // Anfragen welche direkt mit / enden als Anfrage auf die entsprechende index.html (Login) Datei behandeln
  // Sollte eine gültige Session bestehen, dann gleich auf die homepage weiterleiten (Aktuell eingeloggter User versucht die Login page aufzurufen)
  if (request.url.pathname.endsWith("/")) {
    if (!session) {
      request.url.pathname += "index.html";
    } else {
      // Wenn der User gerade eingeloggt ist, dann seine Anfrage an die Index Seite direkt an die homepage weiterleiten
      response.statusCode = 302;
      response.setHeader("Location", config.session.homepage);

      // Anfrage abschließen
      response.end();
      return;
    }
  }

  /*************************************/
  /* ACCESSLIST: BLOCK, LOGIN, GRUPPEN */
  /*************************************/

  // Zugriff auf alle Dateien der Blocklist mit dem Status Code 403 (Forbidden) verbieten
  if (accesslist.block?.includes(request.url.pathname)) {
    response.endWithStatus(403);
    return;
  }

  // Zugriff auf Dateien für Benutzer ohne Anmeldung überprüfen beziehungsweise mit dem Status Code
  // 403 (Forbidden) verbieten
  if (!session) {
    // Wenn eine Allowlist existiert, dann darf ein nicht angemeldeter Benutzer nur auf die Dateien
    // der Allowlist zugreifen und sonst nichts
    if (accesslist.allow instanceof Array) {
      // Zugriffe auf alle Dateien, welche nicht explizit in der Allowlist stehen, verbieten
      if (!accesslist.allow.includes(request.url.pathname)) {
        response.endWithStatus(403);
        return;
      }
    }

    // Existiert keine Allowlist, dann darf ein nicht angemeldeter Benutzer auf die Dateien der
    // `login` Liste nicht zugreifen -> Witerleitung zum Login
    else if (accesslist.login?.includes(request.url.pathname)) {
      response.statusCode = 302;
      response.setHeader("Location", "\\index.html");
      response.end();
    return;
    }
  }

  // Zugriff auf alle Dateien, welche eine bestimmte Gruppe voraussetzen, mit dem Status Code 403
  // (Forbidden) verbieten
  for (const [group, files] of Object.entries(accesslist.groups ?? {})) {
    if (
      files.includes(request.url.pathname) &&
      (!session || !session.groups?.includes(group))
    ) {
      response.endWithStatus(403);
      return;
    }
  }

  /***********************************************/
  /* DATEIEN VOM DATEISYSTEM AUSLESEN UND SENDEN */
  /***********************************************/

  // Alle Filter sind zu diesem Zeitpunkt durchgelaufen, daher versuchen wir nun konkret die ange-
  // forderte Datei zu laden und damit zu antworten
  fs.readFile(`.${request.url.pathname}`, function (error, file) {
    // Fehler beim Zugriff auf die angeforderte Datei mit dem entsprechenden Status Code beantworten
    if (error) {
      // Datei existiert nicht oder würde sich in einem Ordner befinden, welcher nicht existiert,
      // wird mit dem Status Code 404 (Not Found) beantwortet
      if (error.code === "ENOENT" || error.code === "ENOTDIR") {
        response.endWithStatus(404);
        return;
      }

      // Die angeforderte Ressource ist ein Ordner, wofür wir explitit kein Dateianzeige anzeigen
      // wollen, sondern die Anfrage mit dem Status Code 403 (Forbidden) beantworten
      else if (error.code === "EISDIR") {
        response.endWithStatus(403);
        return;
      }

      // Alle anderen aufgetretenen Fehler beantworten wir mit dem Status Code 500 (Internal Server
      // Error), sodass der Benutzer weiß, dass etwas schief gelaufen ist, worauf er keinen Einfluss
      // hat
      response.endWithStatus(500);
      return;
    }

    // Versuchen den Mimetype der angeforderten Datei auf Basis der Dateiendung zu ermitteln und als
    // Header setzen
    const extension = request.url.pathname.split(".").at(-1);
    const mimetype = mimetypes[extension];
    if (mimetype) {
      response.setHeader("Content-Type", mimetype);
    }

    // Die Ausführung von Snippets in der angeforderten Datei auf Basis ausgewählter Dateiendungen
    // erlauben
    if (config.snippets.includes(extension)) {
      // Zunächst die Datei in einen String umwandeln, da sie bisher noch als Buffer vorliegt und
      // dann die darin enthaltenen Snippets ausführen. Zur Ausführung werden einige Objekte des
      // Servers durchgereicht, sodass die Snippets darauf zugreifen können.
      file = executeSnippets(file.toString(), {
        request,
        response,
        session,
        ...templates,
      });
    }

    // Anfrage mit der existierenden Datei beantworten
    response.end(file);
  });
});

// Server auf dem angegebenen Port der Konfigurationsdatei starten
server.listen(config.server.port, function () {
  console.log(`Server wurde auf Port ${config.server.port} gestartet`);
});


src/api/memory.js
"use strict";

const crypto = require("crypto");
const sqlite3 = require("sqlite3").verbose();

function getMemories($userId, data = {}, cb) {
  let db = new sqlite3.Database("data/db.sqlite");

  // Prepare the where clause which searches title, description, locations, activities
  // and persons for the entered search string
  // If no search string was entered then let the searchClauses be empty
  let searchClauses = "";
  if (data.hasOwnProperty("$search") && data.$search.trim() !== "") {
    searchClauses = " AND (entry_date LIKE $search OR title LIKE $search OR description LIKE $search OR locations LIKE $search OR activities LIKE $search OR persons LIKE $search)";
    data.$search = `%${data.$search}%`;
  } else {
    delete data.$search;
  }

  // Prepare the where clause which additionally filters the records to
  // 0 -> no favorites
  // 1 -> favorites
  // all other numbers -> both (delete the favorite filter)
  let favoriteClause = "";
  if (data.hasOwnProperty("$favorite") && (data.$favorite === "0" || data.$favorite === "1")) {
    favoriteClause = " AND favorite = $favorite";
  } else {
    delete data.$favorite;
  }

  // Prepare the where ckause which additionally filters the records to the mood
  // 🚫 or an empty string means no mood filter
  let moodClause = "";
  if (data.hasOwnProperty("$mood") && data.$mood.trim() !== "" && data.$mood !== "🚫") {
    moodClause = " AND mood = $mood";
  } else {
    delete data.$mood;
  }

  // Add the $userId to the data object & delete the getMemories property
  // Otherwise the used and defined wildcards don't match -> error
  data.$userId = $userId;
  delete data.getMemories;

  db.serialize(() => {
    // Create the table tbl_memories if no table exists
    db.run(`CREATE TABLE IF NOT EXISTS "tbl_memories" (
      "uuid"	TEXT NOT NULL UNIQUE,
      "entry_date"	TEXT NOT NULL,
      "title"	TEXT NOT NULL,
      "favorite"	INTEGER NOT NULL,
      "mood"	TEXT,
      "description"	TEXT,
      "url"	TEXT,
      "locations"	TEXT,
      "activities"	TEXT,
      "persons"	TEXT,
      "user_id"	TEXT NOT NULL,
      PRIMARY KEY("uuid")
    );`);

    // Get the rows according to the select statement
    db.all(`SELECT * FROM tbl_memories WHERE user_id = $userId${searchClauses}${favoriteClause}${moodClause} ORDER BY entry_date`, data, function (err, rows) {
      if (err) {
        // send the error message back to the requester
        cb(err.message);
        return;
      }
      const resData = [];

      rows.forEach((row) => {
        resData.push(row);
      });

      // send the complete queried memories back to the requester
      cb(null, resData);
    });
  });

  // close the database connection
  db.close();
}

function insertMemory(userId, data, cb) {
  let db = new sqlite3.Database("data/db.sqlite");
  const uuid = crypto.randomUUID();


  //***********TODO: Convert the base64 code to blob and save it in the database */
  //*********** */
  //***************** */


  // If the uuid is empty, then a new entry in the sqlite db will be done
  if (data.uuid === "") {
    data.uuid = uuid;
  }

  data = Object.values(data);
  data.push(userId);

  // Encode the <> to prevent html and script injections
  data = data.map((entry) => entry.replace(/</g, "&lt;").replace(/>/g, "&gt;"));

  // insert one row into the tbl_memories table
  db.run(`INSERT OR REPLACE INTO tbl_memories VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`, data, function (err) {
    if (err) {
      // send the error message back to the requester
      cb(err.message);
      return;
    }
    // send the last inserted id back to the requester
    cb(null, data.uuid);
  });

  // close the database connection
  db.close();
}

function deleteMemory(userId, data, cb) {
  let db = new sqlite3.Database("data/db.sqlite");

  data = Object.values(data);
  data.push(userId);

  // insert one row into the tbl_memories table
  db.run(`DELETE FROM tbl_memories WHERE uuid=(?) AND user_id=(?)`, data, function (err) {
    if (err) {
      // send the error message back to the requester
      cb(err.message);
      return;
    }
    // send the deleted id back to the requester
    cb(null, data[0]);
  });

  // close the database connection
  db.close();
}

module.exports = {
  getMemories,
  insertMemory,
  deleteMemory
}


src/lib/helpers.js
'use strict';


// Module bereitstellen
module.exports = {
  getBody,
  getBodyParams,
  getCookies,
};


//
// getBody()
// Hilfsfunktion zum schnellen Sammeln des kompletten Request Body.
// <- request: IncomingMessage
// <- callback: Function
function getBody(request, callback) {
  // Eingehende Datenstücke als Buffer sammeln
  let body = Buffer.alloc(0);
  request.on('data', function(chunk) {
    body = Buffer.concat([chunk]);
  });

  // Abgeschlossenen Datenempfang behandeln und Callback Funktion aufrufen
  request.on('end', function() {
    callback(null, body);
  });
}


//
// getBodyParams()
// Hilfsfunktion um schnell mittels Callback Funktion ein echtes Objekt auf Basis des eingegangenen
// Query String formatierten Body zu erhalten.
// <- request: IncomingMessage
// <- callback: Function
function getBodyParams(request, callback) {
  // Zunächst die getBody() Funktion verwenden, um den Body als Text zu erhalten
  getBody(request, function(error, body) {
    // Wenn ein Fehler auftreten ist diesen transparent über die Callback Funktion weitergeben
    if (error) {
      callback(error);
      return;
    }

    // Da kein Fehler aufgetreten ist den Body als URLSearchParams parsen und als echtes Objekt im
    // Aufruf der Callback Funktion übergeben
    callback(null, Object.fromEntries(new URLSearchParams(body.toString())));
  });
}


//
// getCookies()
// Automatische Extraktion aller Cookies aus den Header.
// <- request: IncomingMessage
// -> Object
function getCookies(request) {
  // Prüfen, ob überhaupt Cookies an den Server gesandt wurden
  if (request.headers.cookie) {
    // Cookies wie Query Strings behandeln und als URLSearchParams in ein Objekt konvertieren und
    // zurückgeben
    const cookiesRaw = request.headers.cookie.replaceAll('; ', '&');
    return Object.fromEntries(new URLSearchParams(cookiesRaw));
  }

  // Ohne Cookies geben wir ein leeres Objekt zurück
  return {};
}


src/index.js
"use strict";

const $form = document.querySelector("#frm_login");
const $register = document.querySelector("#btn_register");
const $submit = document.querySelector('[type="submit"]');
const $pswList = document.querySelectorAll('[type="password"]');

const $validation = document.querySelector("#p_validation");

const pswPattern =
 "^(?=.*[a-zA-Z])(?=.*[0-9])(?=.*[!@#$%^&*_=+-]).{6,30}$";
const pswTitle =
  "Min. 6, max. 30 characters, at least one number and one special character (!@#$%^&*_=+-)";

/**********/
/* EVENTS */
/**********/

// When clicking on submit button then we will check some things
// according to which view is selected (login / register) and then
// send the request to the server
// Further we register the xhr eventlistener within the submit eventlistener

$form.addEventListener("submit", (event) => {
  event.preventDefault();
  let isLogin = $form.dataset.login === "" ? true : false;

  // Get the username and the password from the form,
  // create a queryString and a XMLHttpRequest
  let username = event.target[0].value;
  let psw = event.target[1].value;

  const params = new URLSearchParams({"username": username, "password": psw});
  const xhr = new XMLHttpRequest();

  // Listen to the response of the server and check if the
  // login/registration did work
  xhr.addEventListener(
    "load",
    () => {
      // If we get not an status of OK or Forwarded, then inform the user
      // with the proper message according to the current view (login/regiater)
      if (xhr.status !== 200 && xhr.status !== 302) {
          alert(`${isLogin ? 'Login' : 'Registrierung'} fehlgeschlagen!`)
        return;
      }
      window.location.href = xhr.responseURL;
    },
    // Because the xhr load listener is created within the
    // submit listener, we only want to create the listener once
    { once: true }
  );

  // Here we are again in the submit listener

  // If we are in the register view, then make the password validation check
  // before send then the request to the server
  if (!isLogin) {
    if ($validation.dataset.matching === "true") {
      xhr.open("post", "/?register", true);
      xhr.send(params);
      return;
    }
    return;
  }
  xhr.open("post", "/?login", true);
  xhr.send(params);
});

// Set an eventListener to toggle between the login view
// and the register view

$register.addEventListener("click", (event) => {
  event.preventDefault();
  if ($form.dataset.login !== undefined) {
    // Click on Sign Up -> switch to register view from sign in view and set action to ?register

    $submit.value = "Sign Up";
    $register.textContent = "Log In";
    $validation.hidden = false;
    $pswList[0].setAttribute("pattern", pswPattern);
    $pswList[0].setAttribute("title", pswTitle);
  } else {

    // Click on Log In -> switch to login view register view and set action to ?login
    $submit.value = "Log In";
    $register.textContent = "Sign Up";
    $validation.hidden = true;
    $pswList[0].removeAttribute("pattern");
    $pswList[0].removeAttribute("title");
  }

  // Toogle the second password input field and the data-login attribute in frm_login
  $form.toggleAttribute("data-login");
  $pswList[1].toggleAttribute("hidden");
  $pswList[1].toggleAttribute("required");
});

// EventListener registered on the password input fields for validating the repeated password -> Check if the repeated password is the same as the first entered on every key up

for (let $element of $pswList) {
  $element.addEventListener("keyup", () => {
    if ($pswList[0].value === $pswList[1].value) {
      $validation.setAttribute("data-matching", true);
      $validation.textContent = "Matching passwords";
    } else {
      $validation.setAttribute("data-matching", false);
      $validation.textContent = "Not matching passwords";
    }
  });
}
